src/concurrency/transaction_context.cpp

```diff
@@ -80,102 +78,63 @@ void TransactionContext::Init(const size_t thread_id,
-  RWType rw_type = RWType::INVALID;
-
-  auto rw_set_it = rw_set_.find(location);
+  const auto rw_set_it = rw_set_.find(location);
   if (rw_set_it != rw_set_.end()) {
     return rw_set_it->second;
   }
-  return rw_type;
+  return RWType::INVALID;
 }
 
 void TransactionContext::RecordRead(const ItemPointer &location) {
-
+  PELOTON_ASSERT(rw_set_.count(location) == 0 ||
+                 (rw_set_[location] != RWType::DELETE &&
+                  rw_set_[location] != RWType::INS_DEL));
   auto rw_set_it = rw_set_.find(location);
   if (rw_set_it != rw_set_.end()) {
-    UNUSED_ATTRIBUTE RWType rw_type = rw_set_it->second;
-    PELOTON_ASSERT(rw_type != RWType::DELETE && rw_type != RWType::INS_DEL);
     return;
   }
-  rw_set_.insert(rw_set_it, std::make_pair(location, RWType::READ));
+  rw_set_[location] = RWType::READ;
 }
 
 void TransactionContext::RecordReadOwn(const ItemPointer &location) {
-  auto rw_set_it = rw_set_.find(location);
-  if (rw_set_it != rw_set_.end()) {
-    RWType rw_type = rw_set_it->second;
-    PELOTON_ASSERT(rw_type != RWType::DELETE && rw_type != RWType::INS_DEL);
-    if (rw_type == RWType::READ) {
-      rw_set_it->second = RWType::READ_OWN;
-    }
-  } else {
-    rw_set_.insert(rw_set_it, std::make_pair(location, RWType::READ_OWN));
-  }
+  PELOTON_ASSERT(rw_set_.count(location) == 0 ||
+                 (rw_set_[location] != RWType::DELETE &&
+                  rw_set_[location] != RWType::INS_DEL));
+  rw_set_[location] = RWType::READ_OWN;
 }
 
 void TransactionContext::RecordUpdate(const ItemPointer &location) {
+  PELOTON_ASSERT(rw_set_.count(location) == 0 ||
+                 (rw_set_[location] != RWType::DELETE &&
+                  rw_set_[location] != RWType::INS_DEL));
   auto rw_set_it = rw_set_.find(location);
-  if (rw_set_it != rw_set_.end()) {
-    RWType rw_type = rw_set_it->second;
-    if (rw_type == RWType::READ || rw_type == RWType::READ_OWN) {
-      is_written_ = true;
-      rw_set_it->second = RWType::UPDATE;
-    } else if (rw_type == RWType::UPDATE || rw_type == RWType::INSERT) {
-      return;
-    } else {
-      // DELETE or INS_DELETE
-      PELOTON_ASSERT(false);
-    }
-  } else {
-    rw_set_.insert(rw_set_it, std::make_pair(location, RWType::UPDATE));
+  if (rw_set_it != rw_set_.end() && (rw_set_it->second == RWType::READ ||
+                                     rw_set_it->second == RWType::READ_OWN)) {
+    rw_set_it->second = RWType::UPDATE;
   }
 }
 
 void TransactionContext::RecordInsert(const ItemPointer &location) {
-  auto rw_set_it = rw_set_.find(location);
-  if (rw_set_it != rw_set_.end()) {
-    PELOTON_ASSERT(false);
-    return;
-  }
-  rw_set_.insert(rw_set_it, std::make_pair(location, RWType::INSERT));
-  ++insert_count_;
+  PELOTON_ASSERT(rw_set_.count(location) == 0);
+  rw_set_[location] = RWType::INSERT;
 }
 
-bool TransactionContext::RecordDelete(const ItemPointer &location) {
+void TransactionContext::RecordDelete(const ItemPointer &location) {
```

Comment:
Please revert this. 
We used to return `true` in case `location` contained `RWType::INSERT` and `false` otherwise. I believe we should maintain these semantics.

Reply:
That value was never captured in any of the calls.